### ISSUE TITLE
Where to replace code blocks with custom JavaScript

### ISSUE THREAD

**Issue by ggoodman** [wooorm/xdm#9]
Given a function `createCodeSnippetComponent` that transforms a snippet of source code into a JavaScript function representing a React component that renders that code, how and where would I hook into the pipeline to swap the node?

Specifically, if the `createCodeSnippetComponent` returns a string representation of the React component, where might I be able to replace the whole fenced code block with an MDX element representing that code?

**Comment by ChristianMurphy** [comment]
https://github.com/wooorm/xdm#components

**Comment by ggoodman** [comment]
Hi @ChristianMurphy, I'm not sure how I'd be able to achieve my goal with runtime components. I'm hoping to make the swap at compile time so that the resulting MDX component has substituted the whole code block with what I've compiled.

It does raise the question though of how this would behave at runtime. The simple approach would mean that components so-produced wouldn't respect any `<code>` overrides.

**Comment by ChristianMurphy** [comment]
>  I'm not sure how I'd be able to achieve my goal with runtime components. I'm hoping to make the swap at compile time so that the resulting MDX component has substituted the whole code block with what I've compiled.

It sounds like what you're really after is https://reactjs.org/blog/2020/12/21/data-fetching-with-react-server-components.html :slightly_smiling_face: 
But until then you could use a https://github.com/wooorm/xdm#optionsrehypeplugins (maybe, though less likely https://github.com/wooorm/xdm#optionsremarkplugins) to swap the rehype node for what you want.
https://unifiedjs.com/learn offers an intro to writing plugins, and https://github.com/rehypejs/rehype-highlight offers an example of what it could look like (or could be used directly)

**Comment by ggoodman** [comment]
Hi @ChristianMurphy, I don't think that this is a use-case for RSC; I specifically want to do this ahead of time so that there is almost no cost at runtime (whether server or client rendered). This question is about using `xdm`'s `compile()` API.

I'm aware of the fact that there are a number of different ASTs involved and have read the docs on the pipeline. What I'm struggling to figure out is which AST, and therefore type of plugin, I should be looking at. Once I find the right AST, I'm comfortable building a plugin but wouldn't know where to start to figure out the shape of the new JavaScript node to produce so that the rest of the XDM pipeline is unaffected.

In other words, what I think I want to do is swap out a `pre>code` node in one of the intermediate AST's and replace it with the code of a React component. I do not want the code that I inject to get highlighted; it _is_ the highlighting code.

**Comment by wooorm** [comment]
It sounds like something very interesting, but it also sounds like you have some more knowledge on what you're steering for? Do you have some background info on why none of the existing rehype plugins are suitable?

**Comment by ggoodman** [comment]
I am doing some code generation of React components that I hope to use both from the context of `.mdx?` code blocks and special import syntax. Having two distinct code formatting pipelines will result in unnecessary bytes being shipped to the browser and will likely make aligning behaviour and styling very challenging.

You can see a demonstration of the generated code, highlighting the code that generated the code here: https://plnkr.co/edit/MTxvFaFkw5dpJ3eb?open=Code.js&preview.

**Comment by ggoodman** [comment]
It just occurred to me that I can probably regex replace the code blocks before handing the off to xdm.

**Comment by wooorm** [comment]
You seem to want to sort of pre-evaluate javascript? If so, there’s also `recmaPlugins` as an option, which work similar to remark/rehype/etc, but on a javascript syntax tree (https://github.com/syntax-tree/esast). Where you’re given something close to: https://astexplorer.net/#/gist/14cebe29b6ca9ee829aa9563c4906568/20db73a0f0ccec56e50404414cb439dc785716f8.
This is similar to building Babel plugins, but there’s no need to learn an API: it’s a json syntax tree.

If you’re going to precompile content, an alternative is to use HTML, which gzips well, and could also be injected?

> I hope to use both from the context of .mdx? code blocks and special import syntax

As in, they both used in the content, *and* they are exported? If they are exported, how would they be named?

> It just occurred to me that I can probably regex replace the code blocks before handing the off to xdm.

How would that work?

**Comment by wooorm** [comment]
I do believe @ChristianMurphy is right that React Server Components could help here. In that demo there was some pretty smart stuff of what data went down the wire, which might make this pre-evaluate thing superfluous?

**Comment by ggoodman** [comment]
I don't think that RSC are a panacea for doing work like this. The syntax highlighting provided by libraries like [shiki](npm.im/shiki) do quite a bit of work. Notably, then need to spin up a WASM module, compile the textmate syntax definition and use that to tokenize the actual code. Doing this in a RSC would mean delivering very minimal bytes to the browser at the expense of repeated, redundant and expensive work on the server side. It also means adopting an experimental build of React that has strong opinions on the server side. I believe that doing the expensive work once, at build time and working to deliver the minimal bytes to the browser is a better set of trade-offs.

While I have a working approach that compiles to `hast`, it is missing a key element that I'm looking for: the ability to select which lines to render and which ranges to emphasize at _runtime_. Line selection is less relevant to code blocks since you'd typically render every line in the code block. However, having a single implementation for all code snippets seems worth it. Line emphasis would be triggered by extra meta information in the code block meta.

**Comment by ChristianMurphy** [comment]
@ggoodman I agree RCS is not a panacea.
However, I do think it solves *your specific use case*, it supports both server side and client side caching. What you are describing is a server side cache.
It can in addition mix in some dynamic runtime elements with the static content.

**Comment by ChristianMurphy** [comment]
> it is missing a key element that I'm looking for: the ability to select which lines to render and which ranges to emphasize at runtime

There are a few approaches you could take.
As noted above, it still sounds like React Server Components would be a good fit.
An alternative would be to inject react elements for each line, and update them using some shared state (likely `Context`)
Another alternative. you could reach in using `ref` or dom selectors, and modify the content on the client directly.

> Line emphasis would be triggered by extra meta information in the code block meta.

:raised_eyebrow: code block metadata (in the markdown/mdx sense) is static, so why would this need to be dynamic/runtime?

**Comment by ggoodman** [comment]
I think at this point we've strayed a bit from the issue at hand. I feel like there is a bunch of context that will be hard for me to share about _why_ I'm hoping to hook into `xdm` as described. While it's fun discussing alternative approaches, it will be hard for these to make sense for my use-case without spending quite a bit of time sharing the context in which I'm working.

Would there be any ways to get a sort of functional equivalent to a regex replace of code blocks _before_ passing the source through the `acorn` parse?

**Comment by wooorm** [comment]
I posted a suggestion before, using recma/esast: https://github.com/wooorm/xdm/issues/9#issuecomment-779638239, and some other ideas.

> While it's fun discussing alternative approaches

I’m not asking questions because it’s my idea of fun; I do find asking questions the best way to figure out what context you have. Context is important. 

> Would there be any ways to get a sort of functional [...] before passing the source through the acorn parse?

Why should this happen before acorn parse? acorn parsing is the very first thing that happens (at the same time markdown is parsed)

**Comment by ChristianMurphy** [comment]
>  I feel like there is a bunch of context that will be hard for me to share about why I'm hoping to hook into xdm as described. While it's fun discussing alternative approaches, it will be hard for these to make sense for my use-case without spending quite a bit of time sharing the context in which I'm working.

Context is important for answering a question, especially when the missing context prevents you from using standard technologies and approaches.
On top of that, the nestled implications that you don't have time to explain your problem and that your time is more valuable, and that we should guess the right answer by reading between the lines in vague comments is not appreciated.
Your original question was answered in https://github.com/wooorm/xdm/issues/9#issuecomment-779420982.
If you have follow on questions take the time to properly explain them.

**Comment by ggoodman** [comment]
>On top of that, the nestled implications that you don't have time to explain your problem and yet, that we have the time guess the right answer by reading between the lines in vague comments, are not appreciated.

Thanks for calling that out. You two have both been engaging in good faith and I appreciate the dialogue. I will come back with a better explanation of the context. Perhaps with that, we'll be able to come to the best solution across the board and maybe all learn something in the process.

**Comment by wooorm** [comment]
<3

**Comment by ggoodman** [comment]
For a project I'm working on ([Nostaglie](github.com/ggoodman/nostalgie)), I plan to be writing several tutorials on how to interact with the framework and its features. Nostalgie is a next.js-like framework that takes a user's code as input and produces a distributable node server that has some novel data fetching / routing / SSR integrations. Part of the philosophy is that the resulting server should follow best performance practices so that SSR is as efficient as possible. A few strategies are currently in place to help make SSR efficient:

- SSR happens in a different thread / process as the `http.Server` so that the blocking, synchronous work of `ReactDOMServer.renderToString()` is not affecting the server's ability to handle incoming requests in a timely manner. A consequence of this is that since SSR happens in a pool of workers, it is non-trivial to share state between these workers (especially if this state is non-serializable).
- SSR is given a time budget for async work, like data fetching. The initial `ReactDOMServer.renderToString()` will always happen, but subsequent renders may or may not happen, depending on how quickly any async work completes.
- Any work that can reasonably be done at compile-time is pre-computed so that runtime renders have as little remaining work to do. The thinking is that the one-time cost of pre-compilation would be amortized over many many requests.

On the subject of the tutorials, it occurred to me that a really nice way to do this would be to create several example apps that demonstrate what I'm trying to teach in the tutorial. I plan to make those examples serve an additional purpose though; to act as the source of source code snippets embedded in the tutorials. In that way, the examples remain the source of truth. I can test the examples and since I'm just pulling their code at build-time, I wouldn't risk copy/paste errors or worse, sharing invalid / broken code in the tutorials.

With that idea in mind, I thought that it would be interesting to be able to import whole source files as syntax-highlighted components. The same component could be used with different props to present different ranges of the underlying source file. To achieve this, I figure that during client-side render (CSR) or server-side render (SSR), I would need some abstract representation of the lines of these source files. That abstract representation should be as small as possible but contain enough information to add the actual syntax highlighting when rendered.

My solution to this was to precompile the source files into a 2d array of (token + coloring) information (2d because its an array of lines of arrays of tokens). At runtime, a function that could be shared by all instances of the snippet highlighter could take the abstract representation and return a React component. The React component would accept props for selecting the range to present and, optionally, ranges therein to emphasize.

I knew that since Nostalgie is a bundler / compiler at its heart, it would be simple to introduce a new import specifier syntax to import this sort of React component. However, it remains uncertain for me how to align this design with `.md` and `.mdx` files. In such files, code blocks have no underlying file so maybe their contents could be treated as a virtual file. I thought that since I already had the infrastructure in place to produce JavaScript _code_ to represent a syntax-highlighted snippet, I could inject that into the `xdm` pipeline that is already designed to handle JavaScript.

**Comment by ChristianMurphy** [comment]
Taking a step back, leaving `mdx` and `md` out of the picture for a moment.
How would Nostalgie handle this style of syntax highlighting with a plain JS file?

> code blocks have no underlying file so maybe their contents could be treated as a virtual file

Could you expand on what you mean by this?
`xdm.compile` returns a [vFile](https://github.com/vfile/vfile) by default, could that be used in Nostalgie? Or is it looking for something more specific?

**Comment by wooorm** [comment]
@ggoodman a lot to think about, thanks, so more later.

My first idea is also, somewhat similar to Christian is thinking, that this might be a thing for JS in general, and not specific to md/mdx. Could this be done by generating a js file (possible virtually) that represents a component?

**Comment by ggoodman** [comment]
I've mostly hacked it together using the regex replace but in running it over my changelog (where code blocks are often indented under list items) is missing those instances. I've picked up a regex that is designed to avoid capturing indented code blocks since these, in theory should _not_ be escaped under normal circumstances. I guess this part of MDX must be contextual on the current level of indentation... something that a regex will likely always fail at.

Anyway, notwithstanding that hiccup, here is an impression of what it looks like: 
![image](https://user-images.githubusercontent.com/50030/108152654-83527280-70a7-11eb-8b2f-3aeebbcb4943.png)

As you can see, there are a few (totally arbitrary) lines that have been highlighted. The relevant code to trigger this looks like: 

    ```tsx emphasize:2-5
    export interface ClientAuthAuthenticated {
      isAuthenticated: true;
      credentials: {
        audience?: string;
        claims: IdTokenClaims;
        scope: string[];
        user: UserinfoResponse;
        idToken: string;
        accessToken: string;
      };
      loginUrl: string;
      logoutUrl: string;
    }
    ```

The React component to render the pre-compiled token array looks like:

```tsx
import * as React from 'react';
import type { CreateSnippetDataResult } from './parser';

export function createSnippetComponent(parseResult: CreateSnippetDataResult) {
  return function BoundCodeSnippet(props: Omit<CodeSnippetProps, 'parseResult'>) {
    return <CodeSnippet {...props} parseResult={parseResult} />;
  };
}

export interface CodeSnippetProps {
  parseResult: CreateSnippetDataResult;
  fromLine?: number;
  toLine?: number;
  emphasizeRanges?: Array<[fromLine: number, toLine: number]>;
}

export function CodeSnippet(props: CodeSnippetProps) {
  const { bgColor, fgColor, firstLineOffset, language, tokensByLine } = props.parseResult;
  const fromLine = Math.max(0, (props.fromLine || 1) - 1);
  const toLine = Math.max(fromLine + 1, props.toLine || tokensByLine.length);
  const lineSlice = tokensByLine.slice(fromLine, toLine);
  const firstLineNumber = firstLineOffset + fromLine + 1;
  const emphasizeLines = React.useMemo(
    () =>
      (props.emphasizeRanges || []).reduce((lines, [fromLine, toLine]) => {
        for (let i = fromLine; i <= toLine; i++) {
          lines.add(i);
        }

        return lines;
      }, new Set<number>()),
    [props.emphasizeRanges]
  );

  return (
    <pre
      data-lang={language}
      className={emphasizeLines.size ? 'has-emphasis' : undefined}
      style={{
        backgroundColor: bgColor,
        color: fgColor,
      }}
    >
      <code>
        {lineSlice.map((line, lineIdx) => {
          const lineNumber = firstLineNumber + lineIdx;

          return (
            <span
              className={emphasizeLines.has(lineNumber) ? 'has-emphasis' : undefined}
              data-lineno={lineNumber}
              key={lineNumber}
            >
              {line.map(([value, fgColor], tokenIdx) => {
                return (
                  <span key={tokenIdx} style={{ color: fgColor }}>
                    {value}
                  </span>
                );
              })}
              {lineIdx < lineSlice.length - 1 ? '\n' : null}
            </span>
          );
        })}
      </code>
    </pre>
  );
}
```

I've been able to set things up so that a single instance of the `<CodeSnippet>` component is present in the front-end and is re-used for each code snippet.

**Comment by wooorm** [comment]
Hmm your previous post seemed to indicate you wanted to get tutorials from an external file, but this one is from markdown?

About the indent: this mdx parser does not allow indented code, so I’m not sure what’s going on there.

How does that replace you’re using look, and where are you doing it?

**Comment by ggoodman** [comment]
I think I'm finally wrapping my head around this pipeline. Looking at the AST you get in `remarkPlugins`, it strikes me that I might be able to achieve exactly what I was hoping for with access to injecting a remark plugin before `remarkMdx`. That way, I'd be able to make strategic substitutions on the MDAST _before_ the mdx magic kicks in.

The alternative, I think, is to use the existing `remarkPlugins` and inject exactly one `mdxjsEsm` node for the import of the component (`import { CodeSnippet } from 'path/to/runtime/library'`). Constructing this and all the estree structures I see in `.data.estree` seems doable, but unfortunate. Is that needed? Additionally, each `pre` node whose only child is a `code` node would be replaced with an `mdxJsxFlowElement` node. Constructing this seems to be fairly straight forward.

Would you be open to adding the ability to hook into the pipeline before `remarkMdx`? What do you think of the alternative?

**Comment by ChristianMurphy** [comment]
> I think I'm finally wrapping my head around this pipeline. Looking at the AST you get in remarkPlugins, it strikes me that I might be able to achieve exactly what I was hoping for with access to injecting a remark plugin before remarkMdx. That way, I'd be able to make strategic substitutions on the MDAST before the mdx magic kicks in.

Parsing generally happens all at once.
`remark-mdx` adds some additional parse types into `remark` but it still runs as one unit.

> The alternative, I think, is to use the existing remarkPlugins and inject exactly one mdxjsEsm node for the import of the component (import { CodeSnippet } from 'path/to/runtime/library').

That is how most things are approach in unified, through updating the AST.

---

A side note, the work you are doing sounds a lot like a minimal version of https://github.com/pomber/code-surfer (MDX and Gatsby SSR based) and https://github.com/code-hike/codehike (general React version), you may be able to draw code or inspiration from those projects.

**Comment by ggoodman** [comment]
It seems that I've been able to make it work via the more verbose, alternative approach. The code is an absolute disaster area of comments and `as any`s but it _works_. I've created a couple helpers to create the necessary estree nodes for the import of my component and the passing of the 'parse result' as an attribute to the injected `<CodeSnippet>` component. 🎉 🎉 🎉 

Once I have it cleaned up, I'll circle back with how it hangs together.

Here's a snapshot of the edge-case that was breaking my regex. The regex was failing to account for code blocks nested within multi-paragraph list elements. That this is now highlighted shows that the `remark`-based plugin seems to work.

![image](https://user-images.githubusercontent.com/50030/108299795-1bfef600-716d-11eb-8819-03f94227e6eb.png)

**Comment by wooorm** [comment]
> The alternative, I think, is to use the existing remarkPlugins and inject exactly one mdxjsEsm node for the import of the component (import { CodeSnippet } from 'path/to/runtime/library'). Constructing this and all the estree structures I see in .data.estree seems doable, but unfortunate. Is that needed? Additionally, each pre node whose only child is a code node would be replaced with an mdxJsxFlowElement node. Constructing this seems to be fairly straight forward.

You are now getting pretty close to my earlier suggestion: make a recma plugin!
I do not suggest a remark plugin, because what you are doing has more to do with JavaScript than markdown

**Comment by wooorm** [comment]
Going to close this as this question seems answered!

**Comment by ggoodman** [comment]
For anyone following in my internet footsteps, the code that arose from this discussion can be found here: https://github.com/ggoodman/nostalgie/blob/720d2f5f334ffbaf4201fcf628f6f345af15cc0f/src/worker/mdxCompiler.ts#L76-L160.

Thanks for your help and (especially) patience @wooorm and @ChristianMurphy.

**Comment by wooorm** [comment]
@ggoodman very cool to see the whole, a few of my open questions from the conversion now fall into place. This looks nice!
